src/migrations/1722869409448-stripe-payment-intents.ts

```diff
@@ -0,0 +1,191 @@
+    }));
+  }
+
+  public async down(queryRunner: QueryRunner): Promise<void> {
+    await queryRunner.changeColumn('stripe_payment_intent_status', 'stripePaymentIntentId', new TableColumn({
+      name: 'depositId',
+      type: 'integer',
+      isNullable: false,
+    }));
+    await queryRunner.renameTable('stripe_payment_intent_status', 'stripe_deposit_status');
+
+    const depositTable = await queryRunner.getTable(this.DEPOSIT_TABLE_NAME);
+    const foreignKeys = depositTable.foreignKeys.filter((fk) => fk.referencedTableName === this.PAYMENT_INTENT_TABLE_NAME);
+    await queryRunner.dropForeignKeys(depositTable, foreignKeys);
+
+    const paymentIntentTable = await queryRunner.getTable(this.PAYMENT_INTENT_TABLE_NAME);
+    const paymentIntentStatusTable = await queryRunner.getTable('stripe_deposit_status');
+    await queryRunner.dropForeignKeys(paymentIntentStatusTable, paymentIntentStatusTable.foreignKeys);
+    await queryRunner.dropForeignKeys(paymentIntentTable, paymentIntentTable.foreignKeys);
+    await queryRunner.addColumns(this.PAYMENT_INTENT_TABLE_NAME, [
+      new TableColumn(
+        {
+          name: 'toId',
+          type: 'integer',
+          isNullable: true,
+        }),
+      new TableColumn({
+        name: 'transferId',
+        type: 'integer',
+        isNullable: true,
+      }),
+    ]);
+    await queryRunner.createForeignKeys(this.PAYMENT_INTENT_TABLE_NAME, [
+      ...paymentIntentTable.foreignKeys,
+      new TableForeignKey({
+        columnNames: ['toId'],
+        referencedColumnNames: ['id'],
+        referencedTableName: 'user',
+        onDelete: 'RESTRICT',
+      }),
+      new TableForeignKey({
+        columnNames: ['transferId'],
+        referencedColumnNames: ['id'],
+        referencedTableName: 'transfer',
+        onDelete: 'RESTRICT',
+      }),
+    ]);
+    await queryRunner.createForeignKeys(paymentIntentStatusTable, paymentIntentStatusTable.foreignKeys);
+
+    await queryRunner.query(`
+UPDATE ${this.PAYMENT_INTENT_TABLE_NAME}
+SET
+  toId=(SELECT toId FROM ${this.DEPOSIT_TABLE_NAME} WHERE ${this.DEPOSIT_TABLE_NAME}.id = ${this.PAYMENT_INTENT_TABLE_NAME}.id),
+  transferId=(SELECT transferId FROM ${this.DEPOSIT_TABLE_NAME} WHERE ${this.DEPOSIT_TABLE_NAME}.id = ${this.PAYMENT_INTENT_TABLE_NAME}.id)
+`);
+
+    await queryRunner.changeColumn(
+      this.PAYMENT_INTENT_TABLE_NAME,
+      'toId',
+      new TableColumn(
+        {
+          name: 'toId',
+          type: 'integer',
+          isNullable: false,
+        }),
+    );
+    await queryRunner.changeColumn(
+      this.PAYMENT_INTENT_TABLE_NAME,
+      'transferId',
+      new TableColumn({
+        name: 'transferId',
+        type: 'integer',
+        isNullable: true,
+      }),
+    );
+
+    const table = await queryRunner.getTable(this.DEPOSIT_TABLE_NAME);
+    await queryRunner.dropForeignKeys(this.DEPOSIT_TABLE_NAME, table.foreignKeys);
+    await queryRunner.dropTable(this.DEPOSIT_TABLE_NAME);
```

Comment:
I find this quite scary in production, also because this is slightly untested? 

Would it be an idea to add some test cases for this? that you take the suite for the stripe payment intents, revert the migration, add some legacy data, migrate, and run some tests on the new data as seed?

Reply:
I do not actually delete the deposit table. First, I add the deposit-related columns back to the payment intent table and then I merge the two tables together into a single table. I drop the table here, because in the last step, I rename the payment intents table back to the deposit table. I have verified with a locally seeded SQLite database that the migrations work correctly.

Furthermore, I also do not really know how to build a robust test for this. The codebase will evolve and more attributes might be added to these entities. Furthermore, we are going to add more migrations. So how can I build a test case that reverts this single migration, even when we have more migrations and even migrations that change payment intents? And even if we can do that, these tests will probably break, because we write the code for the entities at this point in time, even though they might (and probably will) change.

Reply:
Fair, but then I do suggest to keep this in dev for another deployment cycle (so not include this in the next release) and see if it works in dev.